### PR TITLE
first meaningful Paint metric

### PR DESCRIPTION
**Pull request by paulirish**
Sketching out how we share the fMP metric as described in #26.

Metrics broke out into separate folders, but matching the audit style so far, as they certainly share the inputs approach.

### CODE REVIEW

File: metrics/first-meaningful-paint/audit.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+var DevtoolsTimelineModel = require('devtools-timeline-model');
+
+module.exports = function(data) {
+  // FIXME(paulirish) save the trace to file in debug file for later analysis.
+  var model = new DevtoolsTimelineModel(data.traceContents);
+  var events = model.timelineModel().mainThreadEvents();
+
+  var navStartEvent = events.filter(e => e.name === 'navigationStart');
+  var firstTextPaintEvent = events.filter(e =>
```
Comment (samccone) [inline metrics/first-meaningful-paint/audit.js]: let's extract the filter body into a named function, (easier to test)

File: metrics/first-meaningful-paint/audit.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    e.categoriesString.includes('blink.user_timing') && e.name === 'firstTextPaint'
+  );
+  var duration = firstTextPaintEvent[0].startTime - navStartEvent[0].startTime;
```
Comment (samccone) [inline metrics/first-meaningful-paint/audit.js]: We should put a guard clause here if either firstTextPaintEvent or navStartEvent is empty

File: metrics/first-meaningful-paint/gather.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    return driver.disableCaching()
+      // Fire up the trace.
+      .then(driver.beginTrace)
+
+      // Go to the URL.
+      .then(_ => driver.gotoURL(url, driver.WAIT_FOR_LOAD))
+
+      // Stop the trace, which captures the records.
+      .then(driver.endTrace)
+      .then(traceContents => ({traceContents}));
```
Comment (samccone) [inline metrics/first-meaningful-paint/gather.js]: this work and the work done here -> https://github.com/GoogleChrome/lighthouse/pull/29/files#diff-f0e1982b7f0bb53cbfb9ad52ce2d34f6R45

look like they might be able to be condensed :fountain:

File: package.json [hunk 4]
```diff
@@ -29,6 +29,7 @@
   "dependencies": {
     "cheerio": "^0.19.0",
     "chrome-remote-interface": "^0.11.0",
+    "devtools-timeline-model": "^1.0.14",
```
Comment (samccone) [inline package.json]: I vote that we :lock: this

File: metrics/first-meaningful-paint/gather.js (hunk 3 above)
Comment (paulirish) [inline metrics/first-meaningful-paint/gather.js]: Will rebase on top of brendan's when it lands.

File: metrics/first-meaningful-paint/audit.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+var DevtoolsTimelineModel = require('devtools-timeline-model');
+
+module.exports = function(data) {
+  // FIXME(paulirish) save the trace to file in debug file for later analysis.
```
Comment (brendankenny) [inline metrics/first-meaningful-paint/audit.js]: we should make a real issue for tracking this. I think we're going to want an easy way to save artifacts in general (e.g. for graphing over time a la bigrig)

File: metrics/first-meaningful-paint/audit.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+var DevtoolsTimelineModel = require('devtools-timeline-model');
```
Comment (brendankenny) [inline metrics/first-meaningful-paint/audit.js]: are we going to stomp on each other's webinspector globals at all?

File: metrics/first-meaningful-paint/audit.js [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  var ret = {};
+
+  var navStartEvent = events.filter(e => e.name === 'navigationStart');
+  var firstTextPaintEvent = events.filter(function findFirstTextPaint(e) {
+    return e.categoriesString.includes('blink.user_timing') && e.name === 'firstTextPaint';
+  });
+  if (firstTextPaintEvent.length && navStartEvent.length) {
+    ret.duration = firstTextPaintEvent[0].startTime - navStartEvent[0].startTime;
+  } else {
+    ret.fault = 'First meaningful paint metric not found';
```
Comment (brendankenny) [inline metrics/first-meaningful-paint/audit.js]: we should also probably track an issue for how to better signal "test not applicable" and "test metrics don't yet handle this case"

File: metrics/first-meaningful-paint/audit.js (hunk 6 above)
Comment (paulirish) [inline metrics/first-meaningful-paint/audit.js]: Good question! TBH i have no idea. I'll find out.

File: metrics/first-meaningful-paint/audit.js (hunk 5 above)
Comment (paulirish) [inline metrics/first-meaningful-paint/audit.js]: filed #35

File: metrics/first-meaningful-paint/audit.js (hunk 7 above)
Comment (paulirish) [inline metrics/first-meaningful-paint/audit.js]: filed #36

**Comment by paulirish** [comment]
Unfortunately the global `WebInspector` objects from `devtools-timeline-model` module and brendan's network listener are now conflicting as Brendan foresaw. :/

It'll require some work to sort them out.

**Comment by paulirish** [comment]
`WebInspector` conflicts [finally sorted](https://github.com/paulirish/devtools-timeline-model/pull/6). Good lord I did not enjoy that. :)

**Comment by samccone** [comment]
👌 ace work

**Pull request merged**